File: spikes/fn-hl7-structural-validator-http/src/main/java/com/function/hl7/structure/validator/Function.scala
package com.function.hl7.structure.validator;

import com.microsoft.azure.functions.ExecutionContext;
import com.microsoft.azure.functions.HttpMethod;
import com.microsoft.azure.functions.HttpRequestMessage;
import com.microsoft.azure.functions.HttpResponseMessage;
import com.microsoft.azure.functions.HttpStatus;
import com.microsoft.azure.functions.annotation.AuthorizationLevel;
import com.microsoft.azure.functions.annotation.FunctionName;
import com.microsoft.azure.functions.annotation.HttpTrigger;

import java.util.Optional;

import scala.util.{Try, Failure, Success}
import scala.concurrent.duration._
import java.util.concurrent.TimeUnit

import scala.concurrent._

import cdc.xlr.structurevalidator._;

/**
 * Azure Functions with HTTP Trigger.
 */
class Function {
    /**
     * This function listens at endpoint "/api/HttpExample"
     */
    @FunctionName("HttpExample")
    def run(
        @HttpTrigger(
          name = "req",
          methods = Array(HttpMethod.GET, HttpMethod.POST),
          authLevel = AuthorizationLevel.ANONYMOUS) request: HttpRequestMessage[Optional[String]],
        context: ExecutionContext): HttpResponseMessage = {
        context.getLogger.info("Scala HTTP trigger processed a request.")

        request.getHttpMethod() match {
          case HttpMethod.POST => {

            request.getBody().isPresent match {
              case true => {

                // ASYNC
                val validator = StructureValidatorConc()

                validator.reportJSON(request.getBody().get) match {

                    case Success( report ) => {
                        request.createResponseBuilder(HttpStatus.OK).body(report).build
                    } // .Success 

                    case Failure( err ) => {
                      request.createResponseBuilder(HttpStatus.BAD_REQUEST).body(err.getMessage).build

                    } // .Failure

                } // .Try  

              } // .true

              case _ => request.createResponseBuilder(HttpStatus.BAD_REQUEST).body("empty request body").build
            } // .match

          } // .post

          case _  => request.createResponseBuilder(HttpStatus.OK).body("Use POST with body text HL7 message").build
          
        } // .request
        
    } // .HttpTrigger
}

File: tools/fn-copy-files/src/main/kotlin/gov/cdc/dataexchange/Function.kt
package gov.cdc.dataexchange

import com.microsoft.azure.functions.annotation.*
import com.microsoft.azure.functions.*
import java.util.*

/**
 * Azure Functions with HTTP Trigger to copy items from one Cosmos DB container to another.
 */
class Function {

    companion object {
        private val logger = LoggerFactory.getLogger(Function::class.java.simpleName)
    }
    @FunctionName("copyItems")
    fun run(
        @HttpTrigger(
            name = "req",
            methods = [HttpMethod.GET, HttpMethod.POST],
            authLevel = AuthorizationLevel.FUNCTION,
            route = "copy/{fromDb}/{fromContainer}/{toDb}/{toContainer}"
        ) request: HttpRequestMessage<Optional<String>>,
        @BindingName("fromDb") fromDb: String,
        @BindingName("fromContainer") fromContainer: String,
        @BindingName("toDb") toDb: String,
        @BindingName("toContainer") toContainer: String,
    ): HttpResponseMessage {

        logger.info("Kotlin HTTP trigger processed a request to copy items from $fromDb/$fromContainer to $toDb/$toContainer")

        // Retrieve source and destination endpoints and keys from headers
        val sourceEndpoint = request.headers["source-endpoint"]
        val sourceKey = request.headers["source-key"]
        val destinationEndpoint = request.headers["destination-endpoint"] ?: sourceEndpoint
        val destinationKey = request.headers["destination-key"] ?: sourceKey
        val partitionKeyPath = request.headers["partition-key"] // This should be the path, like "/myPartitionKey"

        // Validate source headers and partition key
        if (sourceEndpoint.isNullOrEmpty() || sourceKey.isNullOrEmpty() || partitionKeyPath.isNullOrEmpty()) {
            return request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                .body("Missing or incorrect source Cosmos DB connection information or partition key in headers.")
                .build()
        }

        // Initialize Cosmos clients for both source and destination containers
        val sourceClient = CosmosClient(
            databaseName = fromDb,
            containerName = fromContainer,
            endpoint = sourceEndpoint,
            key = sourceKey,
            partitionKeyPath = partitionKeyPath,
            isResponseOnWriteEnabled = true
        )

        val destClient = CosmosClient(
            databaseName = toDb,
            containerName = toContainer,
            endpoint = destinationEndpoint,
            key = destinationKey,
            partitionKeyPath = partitionKeyPath,
            isResponseOnWriteEnabled = true
        )

        try {
            // Read all items from the source container
            val items = sourceClient.sqlReadItems("SELECT * FROM c", Map::class.java).collectList().block()

            // Copy each item to the destination container
            items?.forEach { item ->
                destClient.createWithBlocking(item, PartitionKey(item[partitionKeyPath.substring(1)])) // Assume the partition key path starts with '/'
            }

            // Close the clients
            sourceClient.closeClient()
            destClient.closeClient()

            return request.createResponseBuilder(HttpStatus.OK)
                .body("Successfully copied items from $fromDb/$fromContainer to $toDb/$toContainer")
                .build()

        } catch (e: Exception) {
            logger.severe("Exception thrown: ${e.message}")
            return request.createResponseBuilder(HttpStatus.INTERNAL_SERVER_ERROR)
                .body("Error occurred while copying items: ${e.message}")
                .build()
        }
    }
}
